Migrate Groovy to Kotlin

michael-markl opened the pull request:
### Description

This PR
* migrates the groovy build files to kotlin .kts files
  - benefits: One less language with confusing syntax; it's officially supported by Flutter.
  - downsides: Slightly deviates from other/older project setups; modern projects usually use .kts files (that's at least my feeling)
* moves creating & parsing the build config to a separate buildSrc folder (which allows us to use kotlinx.serialization for type safe parsing the build config)
* Adds checks that disallow building an Android app if the build config does not match the specified Flavor
* upgrades Gradle to 8.12  for better support of .kts files, and the Android Gradle plugin (AGP) from version 7.4.0 to 8.8.2.
* specifies target & source compatibility to JDK 17
* removes pub package `fk_user_agent`  as it does not support the newer Gradle version and is generally unmaintained. I replaced the user agent with a manually constructed one that includes the app name & version, the os & device names and versions.
* upgrades `maplibre_gl` and  `sentry_flutter` to support the Gradle version.
* Renames `app.ehrenamtskarte.MainActivity`  to `app.entitlementcard.MainActivity`. (The main dart package should also be renamed from `ehrenamtskarte` to `entitlementcard` at some point)

Feel free to reject the PR if you don't want to migrate; I mainly wanted to test if the `.kts`  files make sense, and in my opinion, they improve the overall build experience slightly (although Gradle itself is still quite confusing^^).


### Side Effects

Likely none.

### Testing

General smoke testing with emphasis on map stuff (as maplibre was upgraded), and on build-config specifics.

### Issues

Relates to #1669

steffenkleinle reviewed (commented):
Nice stuff! LGTM and really glad to see the migration to `.kts` :tada: Sadly while quickly trying to run smoke tests The flutter dev build fails with java heap space problems. I will try investigate more later.
``` log
Launching lib/main.dart on M2101K6G in debug mode...
Running Gradle task 'assembleBayernDebug'...
Selecting debug keystore

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileBayernDebugKotlin'.
> Error while evaluating property 'friendPathsSet$kotlin_gradle_plugin_common' of task ':app:compileBayernDebugKotlin'.
   > Could not resolve all files for configuration ':app:bayernDebugCompileClasspath'.
      > Failed to transform arm64_v8a_debug-1.0.0-235db911ba279722f5e685f38b0ed30fa7e8570a.jar (io.flutter:arm64_v8a_debug:1.0.0-235db911ba279722f5e685f38b0ed30fa7e8570a) to match attributes {artifactType=android-classes-jar, org.gradle.category=library, org.gradle.libraryelements=jar, org.gradle.status=release, org.gradle.usage=java-api}.
         > Execution failed for JetifyTransform: /home/st/.gradle/caches/modules-2/files-2.1/io.flutter/arm64_v8a_debug/1.0.0-235db911ba279722f5e685f38b0ed30fa7e8570a/7bc9cc43ddeab47f507091cee89f1f0e0b601bdd/arm64_v8a_debug-1.0.0-235db911ba279722f5e685f38b0ed30fa7e8570a.jar.
            > Java heap space

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 8s
Error: Gradle task assembleBayernDebug failed with exit code 1
```

michael-markl commented on frontend/android/app/build.gradle.kts:
```diff
@@ -0,0 +1,205 @@
+        if (System.getenv()["KEYSTORE_PATH"] != null) { // Signing with env variables (CI)
+            println("Selected keystore using env variables")
+            create("release") {
+                storeFile = file("file://" + System.getenv()["HOME"] + project.property("KEYSTORE_PATH"))
```
Again, I don't quite know why this was setup as it is, but I don't necessarily want to change this behavior  in this PR. I changed ` System.getenv()["HOME"]`  to `System.getProperty("user.home")` :)

michael-markl commented on frontend/android/build.gradle.kts:
```diff
@@ -0,0 +1,20 @@
+
+rootProject.buildDir = file("../build")
+subprojects {
+    project.buildDir = file("${rootProject.buildDir}/${project.name}")
```
I believe I did not change the behavior in the migration to .kts.
I'm not sure, why you need an absolute path instead of the default path representation (the File class does provide a default `toString` method) 🤔

andrew8er replied:
This works for me now, no idea why.

michael-markl commented on frontend/android/settings.gradle.kts:
```diff
@@ -0,0 +1,22 @@
+pluginManagement {
+    val properties = java.util.Properties()
+    file("local.properties").inputStream().use { properties.load(it) }
```
I am not a particular fan of this suggestion. Makes it harder (at least for me) to know that `load` is called on `properties`.

andrew8er replied:
For me this makes it easier to read, since I try to keep all initialization code around an object to its declaration. From then on, all data is strictly read-only, with very few exceptions. For me, this is essentially the builder pattern that is used in newer Java code. But this is my personal taste.

michael-markl commented on frontend/android/settings.gradle.kts:
```diff
@@ -0,0 +1,22 @@
+pluginManagement {
+    val properties = java.util.Properties()
+    file("local.properties").inputStream().use { properties.load(it) }
+    val flutterSdkPath = properties.getProperty("flutter.sdk")
+    require(flutterSdkPath != null) { "flutter.sdk not set in local.properties" }
```
I am not a huge fan of this either. I don't see a need to wrap this check in a closure.

michael-markl commented:
@andrew8er Thanks for the great review!

> I still (sometimes, about every second time) get an error when syncing the frontend project, still need to investigate the reason here:
> ```
> Build file '/Users/andrewachter/Work/Tür-an-Tür/entitlementcard/frontend/android/app/build.gradle.kts' line: 63
> 
> Command [app-toolbelt, v0, build-config, to-json, bayern, android] at /Users/andrewachter/Work/Tür-an-Tür/entitlementcard/backend failed with status code 1, error: , out: 
> ```

It seems that your working directory is for some weird reason `/Users/andrewachter/Work/Tür-an-Tür/entitlementcard/backend` (the backend folder) Can you explain how you got there / add some repro steps?

> The Gradle API is accesses (project.buildDir) is also deprecated, but I could not find a good replacement right now, and in good Gradle tradition, this is [documented pretty incompletely](https://docs.gradle.org/current/userguide/upgrading_version_8.html#project_builddir)

Yeah, I would stick to this until the default Flutter setup moves away from that.

I also removed all the intellij files that I added. These changes do not really belong to this PR.

@steffenkleinle I never had this error... Nevertheless I increased the memory budget configured gradle.properties. I am not sure if this solves your problem though.

andrew8er commented on frontend/android/app/build.gradle.kts:
```diff
@@ -0,0 +1,205 @@
+        multiDexEnabled = true
+        versionCode = flutter.versionCode()
+        versionName = flutter.versionName()
+        if (project.hasProperty("VERSION_CODE")) {
+            versionCode = (project.property("VERSION_CODE") as String).toInt()
+            versionName = project.property("VERSION_NAME") as String
+        }
```
The let's keep it for now and have separate PR for this. I created a new Flutter project with the AS project generator and this was absent, so it it is probably some legacy code.

andrew8er commented:
> I still (sometimes, about every second time) get an error when syncing the frontend project, still need to investigate the reason here:
>
>```
>Build file '/Users/andrewachter/Work/Tür-an-Tür/entitlementcard/frontend/android/app/build.gradle.kts' line: 63
>
>Command [app-toolbelt, v0, build-config, to-json, bayern, android] at /Users/andrewachter/Work/Tür-an-Tür/entitlementcard/backend failed with status code 1, error: , out: 
>```

The `workdir` is indeed `/Users/andrewachter/Work/Tür-an-Tür/entitlementcard/backend`. This happens at *sync* time, i. e. when I click the *Sync project with Gradle files* button in the IDE. I wonder if this is related to the "Unsupported modules detected" error that happens at the same time.

michael-markl commented:
> The workdir is indeed /Users/andrewachter/Work/Tür-an-Tür/entitlementcard/backend. This happens at sync time, i. e. when I click the Sync project with Gradle files button in the IDE. I wonder if this is related to the "Unsupported modules detected" error that happens at the same time.

I would assume that it is related! Unfortunately, this error does not appear for me. Could you investigate this ? 🤔 

Which directory did you open in Intellij to get this error? The git root or /frontend or /frontend/android?

I usually only open the git root in Intellij and /frontend/android in Android Studio and both seem to work fine for me.

andrew8er commented:
I used to open the whole project (Git root), when opening only the `/fronted` or `/frontend/android`, this error does not occur. I do get the "Unsupported modules detected" error in all three cases though. 

In `/.idea/modules.xml` there is a reference to a non existent file `$PROJECT_DIR$/frontend/android/frontend_android.iml`, but deleting this does not eliminate that error.  

I will investigate this a little later.

andrew8er commented on frontend/android/buildSrc/src/main/kotlin/BuildConfig.kt:
```diff
@@ -0,0 +1,66 @@
+        process.waitFor()
+        val status = process.exitValue()
+        if (status != 0)
+            throw Error("Command $command at $workingDir failed with status code $status!")
```
When openening the `android` directory in AS, this works for me now.

andrew8er reviewed (approved):
I still get the modules error, but otherwise this works for me now.

steffenkleinle reviewed (approved):
Builds successfully now, smoke tested and did not find any issues.
